_source/_posts/2022-04-26-webauthn-java.md

```diff
@@ -0,0 +1,713 @@
+                          @RequestParam String username,
+                          Model model,
+                          HttpSession session) {
+    try {
+        PublicKeyCredential<AuthenticatorAssertionResponse, ClientAssertionExtensionOutputs> pkc;
+        pkc = PublicKeyCredential.parseAssertionResponseJson(credential);
+        AssertionRequest request = (AssertionRequest)session.getAttribute(username);
+        AssertionResult result = relyingParty.finishAssertion(FinishAssertionOptions.builder()
+            .request(request)
+            .response(pkc)
+            .build());
+        if (result.isSuccess()) {
+            model.addAttribute("username", username);
+            return "welcome";
+        } else {
+            return "index";
+        }
+    } catch (IOException e) {
+        throw new RuntimeException("Authentication failed", e);
+    } catch (AssertionFailedException e) {
+        throw new RuntimeException("Authentication failed", e);
+    }
+}
+```
+
+> **NOTE:** These requests require server state to work properly. The authentication request includes a challenge nonce generated as part of the `StartRegistrationOptions` and `AssertionRequest` objects. The reply from the client must include this nonce, so the server uses the session to cache the short-lived challenge by caching the full registration (or assertion) request object.
+
+Information is sent to the browser, where it's encoded and passed to the WebAuthn API as a [`PublicKeyCredential` object](https://developer.mozilla.org/en-US/docs/Web/API/PublicKeyCredential). The response containing all the data needed for verification is sent to the server and used to create or verify a credential.
+
+#### Registration ceremony
+
+To kick off the registration ceremony, the client submits a uniquely identifying piece of information using a `fetch` request—in this case, a username. The server verifies that the username is unique and then creates and saves a 16-byte user handle. Configuration information about the server (like hostname and origin) was included in the `RelyingParty` bean when the server was created. This bean takes a `PublicKeyCredentialRequestOptions` object to collect the user information, and generates a challenge field, a random buffer of at least 16 bytes long. This object is cached on the server to preserve the challenge field, then turned into a JSON string and sent to the client.
+
+Once the information is received by the client, it's parsed into a JavaScript [`PublicKeyCredentialCreationOptions`](https://www.w3.org/TR/webauthn-2/#dictionary-makecredentialoptions) object. Then it is passed into the `navigator.credentials.create()` function as a `publicKey` option.
+
+> **IMPORTANT:** You will need to grab the client code from the example's [GitHub](https://github.com/oktadev/webauthn-java-example/tree/main/src/main/resources) repository.  Copy the `static` and `templates` folders to your project's `src/main/resources` directory.
```

Comment:
Add `wget` (or something similar) commands to this blockquote to download the files with copy/paste.